boards: ikea-tradfri: add support

basilfx opened the pull request:
Last week I found out that the [IKEA TRÅDFRI](http://www.ikea.com/us/en/catalog/categories/departments/lighting/36812/) lighting products and accessories use the Silicon Labs EZR32MG1P microcontroller. Let it be a coincidence that the Thunderboard Sense (#7929) which was merged this week, uses the same CPU!

Now if we look into their line of products, it happens that they use the same CPU module in all(?) of their products: [here](https://www.heise.de/make/artikel/Ikea-Tradfri-Anleitung-fuer-ein-ESP8266-Lampen-Gateway-3598411.html), [here](http://tradfri.blogspot.nl/2017/04/teardown-of-ikea-tradfri-bulb-led1623g12.html) and [here](https://github.com/basilfx/TRADFRI-Hacking). Nonetheless, you can isolate them from the product, and use them for your own ideas, or leave them in and flash a custom firmware. 

The MCU is a ARM Cortex M4 with 256 kB of flash. It's quite packed (including hardware crypto), flexible pin mapping and has an integrated radio which supports Thread, ZigBee Light Link and more. 

It won't surprise me if people buy the cheapest light available (the GU-10 I used costed me 6,99 euro) and upgrade, hack or repurpose them. Let RIOT-OS be one of the first to support them :-)

This PR is a quick port to this new platform. I have more information on the module over [here](https://github.com/basilfx/TRADFRI-Hacking) and a demonstration video [here](https://www.youtube.com/watch?v=yi_Z2WtmdDU). For flashing you need to connect a JTAG/SWD compatible debugger, but I can flash it the exact same way as the Thunderboard Sense when I use SEGGER JLink. You do not even have to remove the module.

basilfx commented:
This will void the warranty of your IKEA product.

miri64 commented:
Is there a radio on-board (and do we support it already)?

basilfx commented:
It has a radio on-board, but we don't support it yet. But now I have more motivation to do so :-)

It would be a nice product/board for demo purposes.

miri64 commented:
> It would be a nice product/board for demo purposes.

Yes definitely! I've been waiting for hacks like this basically since day one of my contribution (but never found the time to do it myself :().

kYc0o commented:
Eeeexcelleeeent!!! I was thinking about these products since I was aware of them, but I'm glad you also investigated them!

I'll try to grab some and hack them with RIOT.

miri64 commented:
I guess I have to guess I have to go to IKEA again (damn it I was saving up for a new PC!!1! :stuck_out_tongue_winking_eye:)

haukepetersen commented:
Awesome!!! My wife will be sooo happy that I am now motivated to go to ikea (besides the usual hot-dog motivation) :-)

Will review do a review right away.

haukepetersen commented on boards/ikea-tradfri/include/board.h:
```diff
@@ -0,0 +1,74 @@
+ * @{
+ */
+#define LED0_PIN            GPIO_PIN(PA, 1)
+#define LED1_PIN            GPIO_PIN(PB, 13)
```
From your video I take that this pin is connected to the FET switching the actual light, right? So what is LED0 connected to?

basilfx replied:
I have to think about this, and see what is useful. In my case, PA1 was also soldered, but I don't know what it does (haven't traced it properly).

Since the same board is used in different light bulbs or accessories, I had this idea of defining different pins on an additional compiler flag. E.g. `IKEA_PRODUCT=ligthbulb` or something like that.

haukepetersen replied:
I see. So I guess lets leave it for now and see what further reverse engineering brings...

basilfx replied:
After closer inspection, it seems that PA1 is only connected for fixing the PCB. It leads to a dead end (missing resistor).

asmaps replied:
Maybe it's connected in the more expensive versions? The versions that can change the white color between warm and cold (not only brightness) have 2 different LED-types for this - cold whites and warm whites. So there's probably one color on each connector?

basilfx replied:
That is definitely possible. In other teardowns you can see that more pins are connected.

haukepetersen commented on boards/ikea-tradfri/include/periph_conf.h:
```diff
@@ -0,0 +1,143 @@
+        .dev = USART1,
+        .mosi_pin = GPIO_PIN(PD, 15),
+        .miso_pin = GPIO_PIN(PD, 14),
+        .clk_pin = GPIO_PIN(PD, 13),
```
Is there anything connected to these pins (or rather: are they accessible)? If not, we might just want to skip the SPI configuration all together for this board?!

basilfx replied:
There is an additional SPI flash that we might want to access.

haukepetersen replied:
Alright, so never mind.

basilfx replied:
I'm not 100% sure, but based on reading the JEDEC ID and Product Identification, it seems that a 2Mbit flash by Integrated Silicon Solution Inc. (part number [IS25LQ020B](http://www.issi.com/WW/pdf/25LQ025B-512B-010B-020B-040B.pdf)) is used.

basilfx commented:
@kaibeckmann That's cool! I can help you if you want!

I am also working on another PR which will improve support for EFM32, based on feedback by @haukepetersen. Hope to create a PR this weekend.

In the mean time, I have updated https://github.com/basilfx/TRADFRI-Hacking and added some work-in-progress schematics and PCB designs for development. I first need another trip to IKEA!

kaibeckmann commented:
@basilfx Yeah that could speed things up!
Regarding the EFM32 support:  I ran in some issues with the radio stuff there. I'll try to sum it up and send you a mail. Furthermore, I started to put the gecko-sdk in an package. At the moment I only pull the radio driver blob from it and use the header files for the RAIL-API. Maybe that could be a used for the rest of the emlib-stuff as well?

haukepetersen commented:
@kaspar030 I know you won't like the duplicated code in this PR either, but I'd say we merge this anyway and wait for the results of #8058 and clean-up/unify around the EFM32-based boards once a structure is decided.

kYc0o commented:
@basilfx is there a way to backup the existing firmware? (in case of we want to come back to the default behaviour).

kYc0o commented:
Just FYI, there are efforts to communicate with these things using [aiocoap](https://github.com/chrysn/aiocoap), you might be interested in [this](https://github.com/chrysn/aiocoap/issues/67) issue.

basilfx commented:
@kYc0o Yes, I've managed to backup a copy of the firmware. I haven't tried to flash it back, but I will try that tonight (just went to IKEA to buy some more lights).

kYc0o commented:
There is [this](https://github.com/ggravlingen/pytradfri) effort too.

basilfx commented:
I can confirm that the light will still work after downloading the binary (256 Kb), erasing the chip, and then uploading the binary again. I did have to repair the light bulb, but I think that is not due to flashing. It even defaulted to last brightness.

![img_0983](https://user-images.githubusercontent.com/815976/32966219-65b6d8b8-cbd9-11e7-86ab-1d38695a4a56.JPG)

miri64 commented:
Nobody dares to merge this awesome feature?

basilfx commented:
I have a small fix for the SPI pins. I have confirmed that it works, because I can read the SPI chip.

kYc0o commented:
> Nobody dares to merge this awesome feature?

It was maybe acked too fast 😉

basilfx commented:
Now also featured on Hackaday: https://hackaday.com/2017/11/17/hacking-the-ikea-tradfri-light-bulb/

kYc0o commented:
Waaaaao!!! Congrats! 😃

miri64 commented:
Whoopwhoop \o/

miri64 commented:
>> Nobody dares to merge this awesome feature?
>
> It was maybe acked too fast :wink:

Can you point me to the comment where this is stated? I couldn't find it. (and why wasn't the review dismissed then :wink:?)

basilfx commented:
If somebody wants to re-ACK my last commit (d710676), than I will squash, rebase and commit when Murdock agrees!

kYc0o commented:
> Can you point me to the comment where this is stated? I couldn't find it. (and why wasn't the review dismissed then 😉?)

I didn't mean too fast in the sense of "bad-acked" but in the sense that several changes came just after it was acked which can also belong to this PR. We could have merged this as it was acked (since it was OK to ack it, everything worked as expected and the comments were addressed), and the modifications and improvements which came afterwards would have been part of another PR.

Thus, nothing to dismiss, nothing to complain, just a personal impression. Waiting a bit more before merge it gave us other features and better understanding.

kYc0o commented on cpu/efr32mg1p/ldscripts/efr32mg1p132f256gm32.ld:
```diff
@@ -0,0 +1,28 @@
+MEMORY
+{
+    rom (rx)    : ORIGIN = 0x00000000, LENGTH = 262144
+    ram (rwx)   : ORIGIN = 0x20000000, LENGTH = 31744
```
What happened with the remaining 1K of RAM?

basilfx replied:
If I recall correctly, 1 KiB is reserved for the radio. I've [double-checked](https://github.com/SiliconLabs/Gecko_SDK/blob/master/platform/Device/SiliconLabs/EFR32MG1P/Source/GCC/efr32mg1p.ld) with the Gecko SDK, but that number is correct :-)

kYc0o replied:
Ok. I suppose that is managed internally since I don't see any specific instruction in the ld you cite on which they make use of the last 1K of RAM. Though it looks a bit strange but if it works let's continue this way.

gebart commented on cpu/efr32mg1p/ldscripts/efr32mg1p132f256gm32.ld:
```diff
@@ -0,0 +1,28 @@
+
+MEMORY
+{
+    rom (rx)    : ORIGIN = 0x00000000, LENGTH = 262144
```
262144 can be written as 256K to make it a little bit more friendly

basilfx replied:
I can change that, but I have another PR coming up that will use the default Cortex-M linker scripts :-)

haukepetersen commented:
re-ACK your SPI fix -> please squash and I will be happy to press the big green button :-)

basilfx commented:
@haukepetersen Go ahead :-)

haukepetersen commented:
My pleasure! -> go

miri64 commented:
:tada:
